### ISSUE TITLE
Problem with using Stress rig file_scan command on ENGINX

### ISSUE THREAD

**Issue by FreddieAkeroyd** [ISISComputingGroup/IBEX#2650]
The relative value commands rstress / rstrain seem to be either sometimes ignored or send incorrect values

ENGINX tried two scrripts in 
\\ndxenginx\c$\scripts\file_scan

David_CPTi_test.txt and David_CPTi_100C_A.txt

The test script did

stress	5.00	10.00	1.00
savestress	0.00	0.00	0.00
rstress	5.00	10.00	1.00
rstress	7.00	10.00	1.00

This created runs 277205, 277206 and 277207

putlog data is:

2017-09-27T22:16:46	Changed PV: IN:ENGINX:INSTRON_01:STRESS:STEP:TIME:SP new=10 old=60 
2017-09-27T22:16:46	Changed PV: IN:ENGINX:INSTRON_01:STRESS:SP new=5 old=5 
2017-09-27T22:16:51	Changed PV: IN:ENGINX:INSTRON_01:MOVE:GO:SP new=1 old=1 
2017-09-27T22:17:00	Changed PV: IN:ENGINX:DAE:PERIOD:SP new=1 old=1 
2017-09-27T22:17:03	Changed PV: IN:ENGINX:CS:RC:SYNC:SP new=1 old=1 
2017-09-27T22:17:05	Changed PV: IN:ENGINX:DAE:ENDSEWAIT new=1 old=1 
2017-09-27T22:17:05	Changed PV: IN:ENGINX:DAE:BEGINRUNEX new=0 old=0 
2017-09-27T22:17:10	Changed PV: IN:ENGINX:DAE:TITLE:SP new=84 old=84 
2017-09-27T22:17:56	Changed PV: IN:ENGINX:INSTRON_01:STRESS:STEP:TIME:SP new=10 old=10 
2017-09-27T22:17:56	Changed PV: IN:ENGINX:INSTRON_01:STRESS:SP new=14.8109 old=5 
2017-09-27T22:18:00	Changed PV: IN:ENGINX:DAE:ENDRUN new=1 old=1 
2017-09-27T22:18:01	Changed PV: IN:ENGINX:INSTRON_01:MOVE:GO:SP new=1 old=1 
2017-09-27T22:18:10	Changed PV: IN:ENGINX:DAE:PERIOD:SP new=1 old=1 
2017-09-27T22:18:15	Changed PV: IN:ENGINX:DAE:ENDSEWAIT new=1 old=1 
2017-09-27T22:18:18	Changed PV: IN:ENGINX:CS:RC:SYNC:SP new=1 old=1 
2017-09-27T22:18:20	Changed PV: IN:ENGINX:DAE:BEGINRUNEX new=0 old=0 
2017-09-27T22:18:25	Changed PV: IN:ENGINX:DAE:TITLE:SP new=84 old=84 
2017-09-27T22:19:11	Changed PV: IN:ENGINX:INSTRON_01:STRESS:STEP:TIME:SP new=10 old=10 
2017-09-27T22:19:11	Changed PV: IN:ENGINX:INSTRON_01:STRESS:SP new=26.5765 old=14.8109 
2017-09-27T22:19:15	Changed PV: IN:ENGINX:DAE:ENDRUN new=1 old=1 
2017-09-27T22:19:16	Changed PV: IN:ENGINX:INSTRON_01:MOVE:GO:SP new=1 old=1 
2017-09-27T22:19:25	Changed PV: IN:ENGINX:DAE:PERIOD:SP new=1 old=1 
2017-09-27T22:19:28	Changed PV: IN:ENGINX:CS:RC:SYNC:SP new=1 old=1 
2017-09-27T22:19:30	Changed PV: IN:ENGINX:DAE:ENDSEWAIT new=1 old=1 
2017-09-27T22:19:35	Changed PV: IN:ENGINX:DAE:BEGINRUNEX new=0 old=0 
2017-09-27T22:19:40	Changed PV: IN:ENGINX:DAE:TITLE:SP new=84 old=84 
2017-09-27T22:20:30	Changed PV: IN:ENGINX:DAE:ENDRUN new=1 old=1 

The stress is set to 5,  14.81 and 26.57 by the script

looking at the stress log files, the stress readback was around 5 by 22:16:59  and was at 14.84 by 22:18:12 so I'm assuming it was at around 5 when savestress was run, so I would expected values of around 10 and 12 to have been sent by the subsequent commands. I guess it could have been 4.81 when savestress was done, so adding 14.81 is adding 2 * 5 ? then 14.81 ->26.57 is around 12 which is 7 + 5 but I'm just playing with numbers. Something not quite right is going on...

**Comment by FreddieAkeroyd** [comment]
Might be worth considering #2647 while working on this

**Comment by GDH-ISIS** [comment]
For easy diagnosis and visibility, would it not be simple and pertinent to alter the stress rig configurations to also record the absolute stress set point, position set point, and strain set point. The parameters could then be viewed by us and the instrument scientists in the log files.

**Comment by FreddieAkeroyd** [comment]
I would probably add them to the instrument archive instead, if the scientists wished we could create blocks for them too.

**Comment by GDH-ISIS** [comment]
Agreed Freddie. I want it to go one step further so they have visibility in the analysis of their data also rather than be tied to IBEX.

**Comment by Tom-Willemsen** [comment]
The problem was that the "relative"ness of the values was being accounted for twice:

In `_process_stressrig_command()` we were doing `_set_relative_value(saved stress + value)`, but within `_set_relative_value` we were also doing `set_value(current_value + new value)`.

I was able to reproduce this problem against the emulator, the fix is to use `_set_stressrig_setpoint` instead of `_set_relative_value` within the rstress/rstrain/rpos commands.

I will email the scientists to let them know a fix is available.

**Comment by Tom-Willemsen** [comment]
We already do `archive(VAL)` on POS:SP:RBV and all the similar channels, so they're already archived and you can view the historical values in the log plotter? Am I missing something extra here @GDH-ISIS @FreddieAkeroyd ?

**Comment by ChrisM-S** [comment]
Am I right in assuming they are simply logging the read back values of these in the blocks into the .NXS files? Hence the diagnostics of the actual set point are just of interest to us when things don’t work??

**Comment by FreddieAkeroyd** [comment]
Unless they create a block explicitly for :SP (or :SP:RBV) they will just get current value in the Nexus file, which is I think what they need. If we add it to the instrument archive, we have it for diagnosis/plotting, which can be useful; however we have writes to :SP via channel access in the putlog already. What we gain by adding :SP is to log writes/changes to :SP from anywhere which might be useful to prove if a setpoint was actually received via a within IOC route. In the current case the existing putlog was good enough for diagnosis, but having :SP in the archive would allow you to easily plot :SP against current value.

**Comment by Tom-Willemsen** [comment]
Script hotfixed, blocks for setpoint readbacks created.
